File: Splunk Interview/236. Lowest Common Ancestor of a Binary Tree.py
class Solution:
    def lowestCommonAncestor(self, root, p, q):

        def recurse_tree(current_node):
            if not current_node:
                return False
            
            if current_node == p or current_node == q:
                return current_node

            left = recurse_tree(current_node.left)
            right = recurse_tree(current_node.right)

            if left and right: return current_node

            return left or right

        return recurse_tree(root)

        # o(n) complexity
        

File: All Interview Package/Splunk Interview/236. Lowest Common Ancestor of a Binary Tree.js
var lowestCommonAncestor = function(root, p, q) {
    function recursion(node) {
        if (!node) return;
        if (node == p || node == q) return node;

        const left = recursion(node.left);
        const right = recursion(node.right);

        if (left && right) return node;
        return left || right;
    }
    return recursion(root);
}
// o(n)


